### ISSUE TITLE
Proposal: No own class for anonymous SJS-defined objects

### ISSUE THREAD

**Issue by sjrd** [scala-js/scala-js#2009]
## Context

#1795 introduced Scala.js-defined JS classes, traits, and objects. And, in particular, specified that anonymous classes extending `js.Any` would be Scala.js-defined by default. For example, in
```scala
@ScalaJSDefined
trait Foo extends js.Object {
  def x: Int
  def y: Any
}

val obj = new js.Object with Foo {
  val x = 5
  val y = "hello"
}
```
the anonymous class is automatically Scala.js-defined.

As currently specified, an actual JavaScript class is created for this anonymous class, so the ES6 semantics look like this:
```javascript
class $anon extends Object {
  constructor() {
    super();
    this.x = 5;
    this.y = "hello";
  }
}

var obj = new $anon();
```
These semantics have two issues:
* They cause a lot of additional code in JS, compared to the "natural" way of writing the above code, which would be `var obj = {x: 5; y: "hello"}`
* As reported in #2006, such objects do not count as "plain objects" because an "invisible" prototype wound its way into the object's prototype chain.

More globally, the semantics are just not what would be *expected* by default by the developer.

## The proposal at large

I propose to refine the specification from #1795 for *anonymous* classes extending `js.Any`. Instead of simply stating that they be Scala.js-defined JS classes by default, we state the following. An `new` expression for an anonymous class extending `js.Any` is *sugar* for instantiating the *parent* class, then adding fields *on the object itself*. There will be no additional class/prototype.

For example, the above Scala.js example "desugars" into the following Scala.js code (I'm intentionally evasive on what `T` and `U` are, at this point):
```scala
val obj = {
  val o: T = new js.Object().asInstanceOf[T]
  o.asInstanceOf[js.Dynamic].x = 5
  o.asInstanceOf[js.Dynamic].y = "hello"
  o.asInstanceOf[U]
}
```
which would, according to existing semantics, eventually become:
```javascript
var o = {}; // new js.Object() produces {}
o.x = 5;
o.y = "hello";
var obj = o;
```
which is immensely closer to the ideal code. In fact, experience suggests that Closure is able to optimize that into:
```javascript
var o = {x: 5, y: "hello"};
```
which *is* the ideal code.

### Type `T` of `o` and type `U` of the result

The types `T` and `U` don't really matter to understand practically the result of the desugaring. For all practical purposes, `T` and `U` can be thought as `js.Dynamic`, with appropriate casts everywhere.

However, we need to carefully craft them to guarantee that the result of the desugaring is still well-typed from scalac's point of view.

* The type `T` of the temporary `val o` is the *self type* of the anonymous class, as seen from within it, but rewritten not to refer to the `$anon` class symbol itself (since it won't exist after desugaring).
* The type `U` of the result is the `tpe` of the scalac `Block` defining the anonymous class.

In practice, `T` and `U` should both look like `js.Object with Foo { val x: Int; val y: String }`, but there might be some weird scalac details. 

### Parent class and interfaces

* Since a Scala.js-defined JS class must inherit from a JS class, there is always a super class to instantiate.
* Since they can only mix in Scala.js-defined JS traits which are guaranteed to be pure interfaces (no concrete term member), there is nothing to actually mix in during the desugaring.
* Super constructor parameters are appropriately forwarded.

Essentially, the "header"
```scala
new ParentClass(arg1, ..., argN) with Trait1 with Trait2
```
desugars into
```scala
val o: T = new ParentClass(arg1, ..., argN).asInstanceOf[T]
```

### Fields

* `val`s and `var`s in the anonymous class desugar to adding the corresponding fields on the object being constructed.
* References to `val`s and `var`s are desugared into reading the fields from the object

```scala
val x = something
var y = x + somethingElse
```
desugars into
```scala
o.x = something
o.y = o.x + somethingElse
```
or, to be precise, as `o` is a `T` and we are past typer:
```scala
o.asInstanceOf[js.Dynamic].updateDynamic("x")(something)
o.asInstanceOf[js.Dynamic].updateDynamic("y")(
    o.asInstanceOf[js.Dynamic].selectDynamic("x") + somethingElse)
```

### Other statements in the class body

Other statements in the class body, i.e., in the constructor of the anonymous class, are desugared in order, appropriately interleaved with the definition of fields.

### References to `this`

References to `this` in the body of the anonymous class desugar into the temporary `val` of the object being constructed, i.e., `o`.

### Methods

`def`s with parentheses desugar into *fields* added on the object, which are `js.ThisFunction` values:
```scala
def m(a: Int, b: Int): Int = a + b + this.x
```
desugars into
```scala
o.m = js.ThisFunction.fromFunction3({ ($this: o.type, a: Int, b: Int) =>
  a + b + $this.asInstanceOf[js.Dynamic].selectDynamic("x")
})
```
Note that, in the body of `def`s, references to `this` are desugared into `$this` (not `o`), and that `$this` has type `o.type`.

### Getters and setters

Getters (`def`s without `()`) and setters (`def`s whose name ends in `_=`) are desugared into JS getters and setters on the object `o` itself, invoking `Object.defineProperty`:
```scala
def foo: Int = x
def foo_=(v: Int): Unit = x = v
```
is desugared into
```scala
js.Dynamic.global.Object.defineProperty(o, "foo", js.Dynamic.literal(
    "get" -> js.ThisFunction.fromFunction1({ ($this: o.type) => $this.x }),
    "set" -> js.ThisFunction.fromFunction2({ ($this: o.type, v: Int) => $this.x = v })
))
```

## Implementation

This desugaring should be implemented as an additional compiler transformation on the scalac `Tree`s. It must happen before `explicitouter`, which comes before `erasure`. I think the best place is to put it between `refchecks` and `uncurry`.

The implementation will be *very* difficult. I do not expect this feature to be shipped any time soon. And I personally will not have the time to implement it. But it seems *possible*, and the spec looks very nice to me.

**Comment by sjrd** [comment]
Comments welcome ^^

**Comment by gzm0** [comment]
I **like** this. Another step forward for Scala.js interop.

One comment on the implementation though: I do not like that we desugar things inside scalac into the ways we *currently* implement Scala.js defined classes. I really would like to see an IR node for this (which then internally can reuse `MethodDef`, etc.). We'll need a magic "forwarder" method for this somewhere in `scala.scalajs.runtime`, but it's fundamentally not any better/worse than `applyDynamic` etc.

Imagine at some point in time ES allows to write:

```javascript
var x = new Foo { x: 1 };
```

Or maybe more realistically, strong mode requires a slightly different way of writing these objects so that they can remain strong objects. We then have a problem.

Further, having a dedicated IR node would actually allow us to emit the optimal resulting code in the first place. WDYT?

**Comment by sjrd** [comment]
> Or maybe more realistically, strong mode requires a slightly different way of writing these objects so that they can remain strong objects. We then have a problem.

Scala.js-defined JS classes are not supposed to be strong. They already allow to do stuff which is not valid in strong mode.

> I do not like that we desugar things inside scalac into the ways we currently implement Scala.js defined classes.

That's why the spec above is explained in terms of desugaring. It is simply *defined* as the above desugaring.

Of course we can argue that it should not be a desugaring, but have its own semantics, which therefore should be moved to the IR. However, I don't see any compelling reason to do so. The IR should be as simple as possible, but not simpler ((c) Albert Einstein). In this case, I think adding an IR node to represent something we can perfectly well write with the current IR is *not* as simple as possible.

Now assuming we *do* want an IR node, we will need a forwarder, as you say. I could see a forwarder having the following shape:
```scala
def newAnonJSObject(instanceFromParent: js.Dynamic, additionalFields: (name, value)*): js.Dynamic = {
  for ((name, value) <- additionalFields)
    instanceFromParent.updateDynamic(name)(value)
  instanceFromParent
}

val obj = new js.Object with Foo {
  val x = 5
  val y = "hello"
}
// becomes:
val obj = newAnonJSObject(new js.Object(),
    "x" -> 5,
    "y" -> "hello"
)
```
The case of properties (getters and setters) is unclear, though. There are also severe issues on how to represent `this` in the right-hand-side of the pairs. For example,
```scala
val obj = new js.Object {
  val x = 5
  val y = x + 5
}
```
would be hard to desugar into such a method call.

That issue left aside, the forwarder could also take a "dictionary" of properties instead:
```scala
def newAnonJSObject(instanceFromParent: js.Dynamic, additionalFields: js.Dynamic): js.Dynamic = {
  for ((name, value) <- additionalFields.asInstanceOf[js.Dictionary[js.Any]])
    instanceFromParent.updateDynamic(name)(value)
  instanceFromParent
}

val obj = new js.Object with Foo {
  val x = 5
  val y = "hello"
}
// becomes:
val obj = newAnonJSObject(new js.Object(), js.Dynamic.literal(
    "x" -> 5,
    "y" -> "hello"
))
```
In that case, the above call would appear as follows in the IR:
```scala
val obj = newAnonJSObject({}, {
  "x": 5,
  "y": "hello"
})
```
With that form, the *optimizer* or the JSDesugaring can intrinsify this method call to produce "better" code if both arguments are `JSObjectConstr`, by fusing them together, effectively producing the ideal code as well: `{x: 5, y: "hello"}`. And if the arguments are not both `JSObjectConstr`, then it would also intrinsify it to inline/unroll to produce the next-to-ideal code in which properties are added after the fact.

So, unless we need special *semantics* for these objects, I don't think an IR node is needed, and therefore is not desirable either.

**Comment by gzm0** [comment]
> Scala.js-defined JS classes are not supposed to be strong. They already allow to do stuff which is not valid in strong mode.

That's exactly the point, they allow you to do stuff which is not valid, but IIUC they don't *force* you. So you can write them such that the VM can optimize them with shadow classes (whether or not the VM is actually in strong mode).

However, I agree that this might be overkill for the moment, since the desugaring we currently have is always and will always be a valid one (unless Scala.js or ES changes class/object model modalities, which will be a major version break anyways). So we are free to replace it later by an IR node if the need arises given that we can live with the fact, that older IR might result in a less good result.

**Comment by gzm0** [comment]
Assigning to 1.0.0 to not forget about this.

**Comment by gzm0** [comment]
Just as a quick info: I have started playing around with this, just to see how far I get ([my feature branch][1]). If anyone takes this up, please ping me to synchronize, as I will move very slowly if at all.

[1]: https://github.com/gzm0/scala-js/tree/no-anon-class

**Comment by gzm0** [comment]
@sjrd just a quick question: I guess your story for generics and overloading resolution is: Do both erasure and overloading resolution manually in this phase. Or did you think of something niftyer?

**Comment by sjrd** [comment]
Erasure shouldn't be needed. Overloading resolution is tricky, and might require some amount of erasure. Unless we manage to pull something off where the generation of overloading resolution is deferred to after erasure? 

Maybe there's a simple compiler API to compute the erasure of a type.

**Comment by gzm0** [comment]
How do you convert this into a lambda without erasure (especially when `tc.ResultType` is also bounded)?

```scala
def foo[T <: Seq[_], TC[T]](x: T)(implicit tc: TC[T]): tc.ResultType = ???
```

I'll think of something how we can move stuff through the pipeline anyways. Maybe having a phantom anonymous class that we discard in GenJSCode is an option? Getting the trees back might be a bit hairy though :S

**Comment by sjrd** [comment]
Arf, higher-kinded type parameters to the method are really nasty, indeed. Yes, you're right, polymorphic methods cannot be transformed without erasure, because Scala does not have polymorphic function types.

**Comment by gzm0** [comment]
How about this: We move the methods (including accessors) on the object into the "companion" of the anonymous class with an additional this argument. We can then desugar the following

```scala
val obj = new js.Object with Foo {
  val x = 5
  def getter[T]: T = ???
  def method(x: String): String = ???
  def method[T <: Seq[_]](x: T): Int = ???
}
```

into

```scala
object $anon {
  def getter[T](this: js.Object with Foo): T = ???
  def method(this: js.Object with Foo, x: String): String = ???
  def method[T <: Seq[_]](this: js.Object with Foo, x: T): Int = ???
}

val obj = {
  val o = new js.Object with Foo
  o.getter = (this) => $anon.getter[Erased{T}](this)
  o.method = (this, x) => { // max number of args here
    // We need to catch this in the backend somehow to do overloading
    method(this, null: String)
    method(this, null: Erased{T})
    // etc.
  }
}
```

This also solves the problem of private methods :)

**Comment by gzm0** [comment]
However, this made me aware of a more fundamental problem: We do not have a way to desugar varargs in this phase. Only `GenJSCode` is able to produce "raw" varargs. So the desugaring for `method` above doesn't really work (i.e. we just need to defer even more work to the `GenJSExport` phase).

**Comment by gzm0** [comment]
Ok. I got it, this is actually much easier than we thought.

In `RewriteAnonJS`, just do the following transformation. Do *not* delete the anonymous class.


```scala
val obj = new js.Object with Foo {
  val x = 5
  def getter[T]: T = ???
  println("foo")
  def method(x: String): String = ???
}
```

Transforms to:

```scala
val obj = {
  val x = new $anon
  x.x = 5
  println("foo")
}
```

Then, in `GenJSCode`, do the following:

* When generating an anonymous, Scala.js defined class, do not generate fields, exports or constructors.
* When encountering `new $anon` and `$anon` is a Scala.js defined class
** Change `new` to invoke the constructor of the super class (this is easy now, since we are past erasure)
** Generate overloads based on symbols (we already do it that way anyways) and pack them into anonymous methods. Add these methods on the object immediately after the new call.
** The next statements will (automatically) be the constructor body.

That's all. Overloading, varargs, erasure, etc. taken care of.

WDYT?

**Comment by sjrd** [comment]
Ouch ... This gets harder and harder. Maybe there simply does not exist a general way to desugar these things :-s And we'll keep finding more and more difficulties.

The mostly-in-back-end idea that you suggest looks interesting. Maybe we can push it further:

* No additional transform
* In the back-end, generate the anon class as usual, but then pull its components apart and "inline" them when we do the `new $anon` (similar to how we pull apart lambdas)

It would solve all the erasure and overloading issues too. It would also more robustly keep semantics of "every else" (everything but the presence of the class).

However, it's unclear what happens to captures and explicit outers.

**Comment by sjrd** [comment]
> However, it's unclear what happens to captures and explicit outers.

Hopefully, all captures and explicit outers become private (or at least non-exposed) fields of the anon function (and not additional params to methods). In that case, when "inlining" we would "only" have to turn private fields into local variables.

**Comment by gzm0** [comment]
> Hopefully, all captures and explicit outers become private (or at least non-exposed) fields

Yes, they do. But that means we must transform everything in the backend (otherwise method signatures wont match). So I'll need to figure out a way how to "find" the constructor of the class again... Then it is actually pretty easy to do the desugaring (and since we are in `GenJSCode`, no dealing with symbols :)).

**Comment by sjrd** [comment]
> how to "find" the constructor of the class again

It's called `"constructor"`. That's how you find it.

**Comment by gzm0** [comment]
That's not what I mean. We need to have the trees of the class constructor at instantiation site. So in pseudocode:

```scala
case New(ident) if isAnonSJSClass(ident.symbol) =>
  // get tree of ctor of ident.symbol
  // rewrite new to be new of super class
  // emit methods
  // emit ctor body
```

All of these steps are easy except for the first one, since we don't know where in our compilation unit these trees are. In the best case, anonymous classes always come before the class that instantiates them. In the worst case, the order is unknown, so we have to reestablish it.

**Comment by sjrd** [comment]
> In the worst case, the order is unknown, so we have to reestablish it.

Yes, we already do something like that for anon function classes:
https://github.com/scala-js/scala-js/blob/v0.6.7/compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala#L204-L216

**Comment by gzm0** [comment]
Nice! I think I can handle this then.

**Comment by gzm0** [comment]
Ok. I have something working. Just a single problem: The following won't work:

```javascript
class Base {
  constructor() {
    this.foo();
  }
}

class Ext extends Base {
  foo() {
    console.log("foo invoked");
  }
}

new Ext
```

In our translation (if `Ext` were anonymous), the super constructor is invoked, before the methods are defined:

```javascript
var o = new Base;
o["foo"] = function() { ... };
```

In ES5, we of course could do the following:

```javascript
var o = Object.create(Base.protoype);
o["foo"] = function() { ... };
Base.call(o);
```

But IIUC, in ES2015 this is not allowed (at least Node.js refuses it).

**Comment by sjrd** [comment]
Indeed, that's not allowed in es 2015. I think we'll have to live with it.
